docs: add Vendure custom fields blog (#1090)
juspay/superposition: docs/blog/2026-07-07-context-aware-vendure-custom-fields.md

> ---

Reviewer: @sauraww this file needs to be mdx for the animations to render

slug: context-aware-vendure-custom-fields

> title: Context-aware Vendure custom fields with Superposition

Reviewer: ```suggestion
title: Context Aware Custom Form Fields in Vendure
```
With superposition is implied, or it can be added to the URL slug for SEO

description: How Superposition adds channel aware resolved values to Vendure custom fields.
tags: [superposition, vendure, commerce, configuration]
---

import AnimationSlide from '@site/src/components/AnimationSlide';
import { VendureOwnershipFlow } from '@site/src/components/animations';

> 

Reviewer: @sauraww you have to add an export in `superposition/docs/src/components/animations/index.ts` for VendureOwnershipFlow

Commerce data is rarely as global as it first appears.

An online store often needs extra product information: a price band, a fulfilment promise, a merchandising label, or a marketplace-specific badge.

> In Vendure, custom fields are a natural place to store this data because they let teams extend products and product variants directly.

Reviewer: Vendure can link to Vendure's website. One thing that would be helpful would be to add a few lines about what Vendure is
```suggestion
In [Vendure](https://vendure.io), custom fields are a natural place to store this data because they let teams extend products and product variants directly.
```

> But some values are "default" values.

Reviewer: This seems out of place

The effective value can change by channel, currency, market, customer group, product, variant, or a rollout state.
That is where things usually start to get messy: more custom fields, more conditional code, more duplicated entities, more operational risk.

The [Superposition Vendure plugin](https://github.com/juspay/vendure-plugin-superposition) takes a cleaner approach:

```text
Vendure owns the stored fallback.
Superposition owns the context-aware effective value
```

<AnimationSlide width="100%" height={420} mode="auto" loop={false}>
    <VendureOwnershipFlow />
</AnimationSlide>

Vendure remains the system of record for the catalog. Superposition adds a runtime configuration layer that can resolve the right value for the current business context.

> <!-- truncate -->

Reviewer: truncate removes everything below this

<img width="1080" height="906" alt="Image" src="https://github.com/user-attachments/assets/37c0094e-a8ab-415d-bd70-5ca5ec9b10b4" />

## Why Superposition fits this problem

> Superposition starts from a simple idea: configuration should be resolved in context. Instead of spreading rules across environment variables, hardcoded conditionals , or one-off feature flags, it lets teams define typed defaults and resolve the right value for each context in a cascading way.

Reviewer: ```suggestion
Superposition starts from a simple idea: configuration should be resolved with context. Instead of spreading rules across environment variables, hardcoded conditionals , or one-off feature flags, it lets teams define typed defaults and resolve the right value for each context in a cascading way.
```

> The important part is that the catalog does not need to become more complex just because the business rules become more specific. Vendure keeps the base product data understandable, and Superposition resolves the effective value at read time.

Reviewer: ```suggestion
The important part is that the catalog does not need to become more complex just because the business rules become more specific. Vendure keeps the base product data understandable, and Superposition resolves the effective value at when required.
```
Is this better?

## A concrete example

Suppose a merchant wants to store a `bulkPrice` for each product variant. In Vendure, this can be added as a custom field on `ProductVariant`.

```ts
SuperpositionPlugin.init({
    entities: {
        ProductVariant: [{ name: 'bulkPrice', type: 'int', nullable: true }],
    },
});
```

Vendure stores the fallback value:

```text
ProductVariant.customFields.bulkPrice = 800
```

That value is still useful. It is the base price that applies when no more specific rule is present.

Now suppose the same merchant wants a different bulk price for the B2B channel in INR. That value does not need to become another Vendure custom field. It can be stored as a contextual override in Superposition:

> ```text

Reviewer: We should write this like cac toml, just to keep the format uniform across docs and blog
